src/libraries/System.Security.Cryptography.Encoding/tests/Cbor.Tests/CoseKeyHelpers.cs

```diff
@@ -0,0 +1,261 @@
+            }
+
+            // run these first to perform necessary validation
+            (CoseKeyType kty, CoseCrvId crv) = MapECCurveToCoseKtyAndCrv(ecParams.Curve);
+            CoseKeyAlgorithm alg = MapHashAlgorithmNameToCoseKeyAlg(algorithmName);
+
+            // begin writing CBOR object
+            writer.WriteStartMap();
+
+            WriteCoseKeyLabel(writer, CoseKeyLabel.Kty);
+            writer.WriteInt32((int)kty);
+
+            WriteCoseKeyLabel(writer, CoseKeyLabel.Alg);
+            writer.WriteInt32((int)alg);
+
+            WriteCoseKeyLabel(writer, CoseKeyLabel.EcCrv);
+            writer.WriteInt32((int)crv);
+
+            WriteCoseKeyLabel(writer, CoseKeyLabel.EcX);
+            writer.WriteByteString(ecParams.Q.X);
+
+            WriteCoseKeyLabel(writer, CoseKeyLabel.EcY);
+            writer.WriteByteString(ecParams.Q.Y);
+
+            writer.WriteEndMap();
+
+            static (CoseKeyType kty, CoseCrvId crv) MapECCurveToCoseKtyAndCrv(ECCurve curve)
+            {
+                if (!curve.IsNamed)
+                {
+                    throw new ArgumentException("Only named curves supported in EC COSE keys.", nameof(curve));
+                }
+
+                if (MatchesOid(ECCurve.NamedCurves.nistP256))
+                {
+                    return (CoseKeyType.EC2, CoseCrvId.P256);
+                }
+
+                if (MatchesOid(ECCurve.NamedCurves.nistP384))
+                {
+                    return (CoseKeyType.EC2, CoseCrvId.P384);
+                }
+
+                if (MatchesOid(ECCurve.NamedCurves.nistP521))
+                {
+                    return (CoseKeyType.EC2, CoseCrvId.P521);
+                }
+
+                throw new ArgumentException("Unrecognized named curve", curve.Oid.Value);
+
+                bool MatchesOid(ECCurve namedCurve) => curve.Oid.Value == namedCurve.Oid.Value;
+            }
+
+            static CoseKeyAlgorithm MapHashAlgorithmNameToCoseKeyAlg(HashAlgorithmName name)
+            {
+                if (MatchesName(HashAlgorithmName.SHA256))
+                {
+                    return CoseKeyAlgorithm.ES256;
+                }
+
+                if (MatchesName(HashAlgorithmName.SHA384))
+                {
+                    return CoseKeyAlgorithm.ES384;
+                }
+
+                if (MatchesName(HashAlgorithmName.SHA512))
+                {
+                    return CoseKeyAlgorithm.ES512;
+                }
+
+                throw new ArgumentException("Unrecognized hash algorithm name.", nameof(HashAlgorithmName));
+
+                bool MatchesName(HashAlgorithmName candidate) => name.Name == candidate.Name;
+            }
+        }
+
+        public static (ECParameters, HashAlgorithmName) ReadECParametersAsCosePublicKey(CborReader reader)
+        {
+            Debug.Assert(reader.ConformanceLevel == CborConformanceLevel.Ctap2Canonical);
```

Comment:
should this method not be public? Otherwise you should not validate with asserts since they will disappear in release builds

Reply:
Good point, I'll make private.